pkg/scheduler/partition.go

```diff
@@ -1433,6 +1433,13 @@ func (pc *PartitionContext) cleanupExpiredApps() {
 		delete(pc.rejectedApplications, app.ApplicationID)
 		pc.Unlock()
 	}
+	for k, app := range pc.completedApplications {
+		pc.Lock()
+		if app.IsExpired() {
+			delete(pc.completedApplications, k)
+		}
+		pc.Unlock()
+	}
```

Comment:
This will cause a data race. The completedApplications map can be manipulated from a different go routine. Maps are not go routine safe. You will need to introduce a `pc.GetCompletedAppsByState()` method to do the filtering under a lock like we do for rejected apps.
The delete will still need to be within the lock.

Reply:
Thanks Wilfred, addressed in [209b0dc](https://github.com/apache/yunikorn-core/pull/463/commits/209b0dc7d4822d3e8cea3d3571eb13121ba39d17)